### PR TITLE
Action button refactor/rework: Enhanced Dragging

### PR DESCRIPTION
**Pull request by LemonInTheDark**
## About The Pull Request

I noticed a lot of strange and un-intuitive behavior in action buttons, and got stung by the bloat bug. Damn it hug #58027
I'll do my best to explain what I've changed and why, might get a bit long.
If you want a better idea, read the commits. Most of em are pretty solid, if long.

Whelp. Here we go. 

### How do action buttons currently work

All action buttons are draggable, to any place on the screen. They're held in an actions list on the player's mob. 
Their location in this list determines their position on the top of the screen. If one is dragged away from the top, its position in the list is "saved". This looks really bad.
If two buttons are dragged over each other, their positions swap. (inside the actions list too) 
If a button is shift clicked, it is brought back to the position it started at.
If the action collapse button that you likely just mentally edit out is *alt* clicked, it resets the position of all action buttons on the screen.
If an action is ctrl clicked, it is "locked". This prevents any future position changes, and also enables a saving feature. With this saving feature, locked button positions persist between rounds. So your first o2 canister will always start where you saved it, etc.
Actions and buttons are a one to one link. While there is functionality to share action buttons between two players, this means showing the same object to both. So one player can move a button on another's screen. Horrendous.
This also makes code that modifies properties of the screen object itself very clunky. 

#### Why is this bad

A: None knew pretty much any of this information. It is actually documented, just in a horribly formatted screen tip on the collapse button, you know the one we all mentally delete from the hud.
B: None of this is intuitive. Dragging buttons makes the hud look much worse, and you get no feedback that you even can drag them. Depressing
C: We use actions to make new options clear to the player. This means players can have a **lot** of action buttons on the hud. This gets cluttery
D: The collapse button is useless. It lets you clear your screen if someone like me fucks up and gives you 2000 actions, but outside of that it just hides all information from you. You never want to see none of your action buttons, just a filtered list of them.
E: On a technical level, they're quite messy, and not fully functionally complete. This is depressing.

### What I've done

Assuming the above to be true, how do we fix them?
Well first I'm going to go over everything I changed, including links to major commits. I'll then describe the finished product, and why I made the decisions I did.

Oh and I've moved some of the more niche or technical discussion to dropdowns. Hopefully this makes finding the major functional changes easier

Adds helper procs for turning screen_loc strings into more manageable arrays. This doesn't fully support all of the screen_loc spec, but it's enough for what I'm doing. (f54865f840634829b5751f128b657d022bc4e4c6)

Uses these helper procs to improve existing code (6273b931aa52d4701bf897c1ffe0f30f403c4af4)

Fixes an issue with tooltip code itself. If you tried to hold down a mouse button while dragging onto a tooltip enabled object, it would silently fail. The js made assumptions about the order args came in, which broke when buttons were held down (e0e42f621771ad742112ef1255129d2177fd8532)

Adds a signal linked to /client/Click(). Surprised we didn't have this before honestly (c491a4aa8cb9d349ae26f25ecfed420bde43adbb)

Makes /client/MouseDrag() return parent. If we don't do this, any overrides of MouseDrag will never actually be called (2190b2acc549cb3f7a379d4fa78ddba8ed9031a3)

<details><summary>Refactors how action buttons work under the hood (53ccce27ba33c7bfe29ace67883ff539d35901ff)</summary>
Basically, rather then generating one button per action, we generate one
button per viewer

This means anyone "sharing" actions will be able to independantly move
them about.

As a consequence of this, we generate the button when we gain a viewer,
rather then when an action is created. This means we can just remove
redundant code that simply updated the viewer's position.

Another side effect is we need to try and update more then just one
button. I've rewrote code to properly support this.

There's some other stuff going on here, which I'll get into later. It's
mostly just better management of button position. You'll see.
</details>

Starts to change button behavior, more cleanup

Changes the mouse cursor when an action button is dragged. Hopefully
this makes moving things feel less like an accident, and makes you doing
it more clear

Removes the moved and locked vars. This will be more relevant later, but
for now:

Moved exists as a sort of budget "We've been dragged" variable. We can
handle this more cleanly, and the movable type doesn't care about it

Locked is a very old variable that is also not something that the
movable type "owns". It's more an action button thing that's been moved
down.
It exists so an action can be locked in place, and in that locking, be
treated as a "saved location"
(21e20fc3478ceeeca4d99015db9fe4a3b2376c8c)

Because I've nuked move, we don't need to directly set our button's
position. We can use the default_button_position var instead. This is
quite handy.

Please ignore `position_action`, I will explain that later
(83e265ef5b051558b7b84a5630e416ec89d09a98)

Removes the buttons locked pref

It was another obscure part of action buttons, basically do buttons
start "locked" or not. See previous discussion of locked
(b58b1bdb2eeab353e94001fbad8b9e1a094f25b7)

Major rework starts here

Alright. Sorry for this, this is where me not commiting regularly starts
to suck. I'll do my best though.

Rather then figuring out an action button's position via a combination
of the moved and ordered vars, we use a separate location var to store
one of a few defines. This makes life later much easier.

Adds tooltip support for dragging action buttons. The way the tooltip
just froze in place when dragging really bugged me, and lead to some
nasty visual artifacts.
This is a bit messy because the drag procs are horrible, but it's
workable

Dropping a button on another button will no longer swap their positions
Behavior instead depends on the target button.

If it's a part of a group (A concept I will explain later) the dragged
button is simply inserted before it in the group's list.

If it's floating on the general hud, we instead position the dragged
button to its right. There's extra logic here to ensure buttons will
never overflow the screen, but I'll get into that later.

Alright. That's most of the refactoring. Time for the larger behavior
changes.

Adds a button palette. This is a separate dropdown that renders
underneath buttons.

image

The idea is to allow for a conceptual separation between "important"
buttons and the ones that end up cluttering the screen.

You can click on the dropdown to open it, then any later clicks that
don't involve actions in some way will autoclose it.

My goal is to come up with an alternative for the action button that
just acted as a way to hide all buttons on screen. Not convinced it saw
much use.

As a side effect of removing that, I've moved its tooltip stuff to the
palette. I've properly formatted it, so hopefully it's easier to read
then the jumble that we used to have.

(You can alt click the palette button to reset all button positions)

Oh and the palette can scroll, since as you'll see later it has a
limited size.
image

Moving on from that, I've added what amounts to action landing buttons.
These allow buttons to rejoin groups, or be positioned at the end of a
line of buttons.
image

They've got a 32x32 hitbox, and only show up when dragging. Hopefully
this makes the system more clear just by dragging an action.

Oh and I've changed how button position updating works. The old system
of calling update_action_buttons on mob every time an action button
changes position is gone, mostly because I've setup more robust
grouping. Will discuss when I get to huds

(0d1e93ffbd5b0915b09fbc0c349410a125bac4cc)

<details><summary>Adds the backbone behind action button position changes (94133bda4d28b3ccfb04200bb0fe3e354fc4db27)</summary>

Alright, so I've showed you how actions think about positioning, but how
does it actually work.

The main thing we're doing here is grouping action buttons more
expansively.

The /datum/action_group datum "owns" a set of action button screen
objects.
It manages their position in a list, and reacts to scrolling.
Also manages the existence of landing objects, and other such things.

Subtypes of the action group include /listed, which stores the action
buttons that live in the top bar of the screen, and /palette, which
holds the buttons stored in the action palette

The palette datum also manages the position of all other palette related
things, like the dropdown button, and scroll buttons.

This dynamically reacts to the height of the list group.

Floating action buttons require much less cohesiveness. I could
probably handle them as a datum too, but honestly since they have no
relation to each other, a list is more suitable.

In addition to all this, action groups will automatically resize if
their rows overflow the screen space. The same can be said for floating
action buttons.

As things currently stand, you can cause byond to display the "fucky"
part of the rendering space just by putting an action button to the
right in widescreen, and then switching to 15x15 space. S dumb. They'll
now relocate to prevent fuckyness.

Kind of on the same topic, floating action buttons can no longer "cut"
the border of the screen. No matter where you try and place one, it will
always remain fully visible, won't get halfway cutoff by the screen
border
</details>

Moves hud defines to the global folder, safer this way  (726011744616d32b8dbd64a5bd1b23182cfea4d4)

Adds color changing to the palette button, giving some heads up for buttons being inserted into the palette automatically
image
image
Ensures a landing button is always shown, even if it needs to break the
max row rule
Makes palettes auto contract if they have no buttons inside them
Prevents palettes from being opened if they have no buttons inside them
(f9417f3861e76676aef83d62a97324de05a98100)

### How it looks

https://user-images.githubusercontent.com/58055496/155839907-a9f38e82-777a-4126-86f1-f687fbd4768d.mp4

## Why It's Good For The Game

Players have more control over the clutter on their screen. 
Buttons are available, but not in the way,
Since any player move of a button saves it, any lack of clarity in the way buttons work will be forced out by buttons not just resetting when a new game starts.
We don't overlap any existing screen elements, unless the upper button list gets really long.
The code is much less crummy (I think, may have made it worse it's hard for me to judge my own work)

If it ends up not being as usable as I'd like, I'll rip out the existing changes and just implement the qol and backend stuff. I think it's worth doing though.

## Changelog

:cl:
add: Expanded heavily on action buttons
add: Adds an action button dropdown that sits just under the normal list in the top left. You can drag new buttons onto it to insert them. Click on it to show its contents, do what you want to do, then click again anywhere to contract it. Alt click it to reset all button positions
add: Action buttons will now remember their position between rounds. So if you really like your flashlight right next to your player for some reason, we support that now 
add: When you start to drag an action button, docking ports will appear in places that it can be inserted into. (Outside of just floating somewhere on your screen of course)
del: Removed action button locking, and the associated preference. I'm reasonably sure literally none uses this, but if you do hit me up
qol: Dragging an action button will now give you an outline of its size around your cursor
fix: You can no longer cause the screen to expand by putting an action button on the edge of widescreen, and then resizing to standard.
refactor: Refactors action and button code significantly. lots of little things.
/:cl:

### CODE REVIEW

**Comment by Ryusenshu** [comment]
Looks pretty great
I was always annoyed by the antag panel being there, same with pda light switch
Those are things you often just look at once or use very rarely, with no point being up on screen forever

**Comment by Ryusenshu** [comment]
Only issue i have with cult spell bar is the location really
Once you have more stuff in your inventory it overlays

**Comment by AffectedArc07** [comment]
poggers

File: code/__DEFINES/hud.dm [hunk 1]
```diff
@@ -14,3 +14,202 @@
...
+	In addition, the keywords NORTH, SOUTH, EAST, WEST and CENTER can be used to represent their respective
+	screen borders. NORTH-1, for example, is the row just below the upper edge. Useful if you want your
+	UI to scale with screen size.
+
+	The size of the user's screen is defined by client.view (indirectly by world.view), in our case "15x15".
+	Therefore, the top right corner (except during admin shenanigans) is at "15,15"
+*/
+
+/proc/ui_hand_position(i) //values based on old hand ui positions (CENTER:-/+16,SOUTH:5)
+	var/x_off = -(!(i % 2))
```
Comment (Mothblocks) [inline code/__DEFINES/hud.dm:38]: bruh

Comment (Mothblocks) [reply code/__DEFINES/hud.dm:38]:
```suggestion
	var/x_off = i % 2 ? 0 : -1
```

get outta here with your clever code

Comment (Mothblocks) [reply code/__DEFINES/hud.dm:38]: oh you didnt write this did you, look how silly it is

Comment (LemonInTheDark) [reply code/__DEFINES/hud.dm:38]: dude that's a heck of a code snippet

**Comment by o0cyann0o** [comment]
FUCK yes brother

**Comment by LemonInTheDark** [comment]
You guys are all way too positive, was expecting some good old fashioned hate comments

**Review by Twaticus — APPROVED** [review]
ive no idea whats happening in this pr but pretty buttons : )

**Comment by Astrogem2** [comment]
lemon, can it be vertical please, I want it to be like before the update

**Comment by LemonInTheDark** [comment]
It won't make sense when the dropdown is displayed

**Comment by Serxule** [comment]
literally just takes up more space when unhidden and a permanent bar thats harder to click than a button

**Comment by LemonInTheDark** [comment]
Have some more work I want to do here. Change the name to stop confusing people, and make it auto open when a button is dragged onto it

> takes up more space when hidden, and unhidden, harder to use, overall worse in every way than the old design.

Did you actually use the old design? As far as space goes, I promise you it will just melt into your subconscious like the old one did

**Comment by LemonInTheDark** [comment]
Decided against auto opening, since it to some extent defeats the point. The glint should help, alongside a non crummy tooltip

**Comment by dragomagol** [comment]
image

Weird side effect, it shows for action buttons on the lobby screen

**Comment by TheSilverNuke** [comment]
> You guys are all way too positive, was expecting some good old fashioned hate comments

IDK why anyone would dislike this PR. Action buttons have been a pain in the ass for years now. Thank you for fixing it.

File: code/_onclick/hud/action_button.dm [hunk 2]
```diff
@@ -1,65 +1,55 @@
+// Todo:
+// Add proper garbage collection
+// Cry yourself to sleep over the existence of hud code
```
Comment (Watermelon914) [inline code/_onclick/hud/action_button.dm:3]: is this still needed to be done, or is the pr ready now?

**Comment by LemonInTheDark** [comment]
> Weird side effect, it shows for action buttons on the lobby screen

Ye, not sure what I'm gonna do about that. it might in theory be useful, but it's like, the title screen

File: code/_onclick/hud/action_button.dm (hunk 2 above)
Comment (LemonInTheDark) [reply code/_onclick/hud/action_button.dm:3]: Reasonably sure it's done, will test it tho

Comment (LemonInTheDark) [reply code/_onclick/hud/action_button.dm]: Did some testing, I think we're good now?

**Comment by LemonInTheDark** [comment]
This is ready

**Pull request merged**